lib/ClangImporter/ClangSourceBufferImporter.cpp

```diff
@@ -29,69 +30,105 @@ static SourceLoc findEndOfLine(SourceManager &SM, SourceLoc loc,
-    mirrorID = mirrorIter->second;
+    mirrorID = mirrorIter->second.FileID;
+    allDirectives = mirrorIter->second.Complete;
   } else {
     std::unique_ptr<llvm::MemoryBuffer> mirrorBuffer{
-      llvm::MemoryBuffer::getMemBuffer(buffer.getBuffer(),
-                                       buffer.getBufferIdentifier(),
+      llvm::MemoryBuffer::getMemBuffer(buffer->getBuffer(),
+                                       buffer->getBufferIdentifier(),
                                        /*RequiresNullTerminator=*/true)
     };
     mirrorID = swiftSourceManager.addNewSourceBuffer(std::move(mirrorBuffer));
-    mirroredBuffers[buffer.getBufferStart()] = mirrorID;
+    mirroredBuffers[buffer->getBufferStart()] = { mirrorID, !forDiagnostics };
+
+    // Make sure to keep a reference to the underlying manager
+    if (clangSourceManagers.insert(&clangSourceManager).second) {
+      clangSourceManagerRefs.emplace_back(&clangSourceManager);
+    }
   }
-  loc = swiftSourceManager.getLocForOffset(mirrorID, decomposedLoc.second);
 
-  auto presumedLoc = clangSrcMgr.getPresumedLoc(clangLoc);
-  if (!presumedLoc.getFilename())
+  SourceLoc loc = swiftSourceManager.getLocForOffset(mirrorID, offset);
+  if (allDirectives)
+    return swiftSourceManager.getLocForOffset(mirrorID, offset);
+
+  // Now add any line directives, if any. Note that the line table will be
+  // incomplete if a module is broken and diagnostics are output, but otherwise
+  // we'll be reading from the serialized module and hence have the full line
+  // table.
+  //
+  // Always add the enclosing line directive if importing a source location for
+  // a diagnostic, but otherwise assume we have the full line table and add
+  // all directives at once to avoid adding a virtual file for every location.
+
+  bool Invalid = false;
+  const clang::SrcMgr::SLocEntry &Entry =
+      clangSourceManager.getSLocEntry(clangFileID, &Invalid);
+  if (Invalid || !Entry.isFile())
     return loc;
-  if (presumedLoc.getLine() == 0)
-    return SourceLoc();
 
-  unsigned bufferLineNumber =
-    clangSrcMgr.getLineNumber(decomposedLoc.first, decomposedLoc.second);
-
-  StringRef presumedFile = presumedLoc.getFilename();
-  SourceLoc startOfLine = loc.getAdvancedLoc(-presumedLoc.getColumn() + 1);
-
-  // FIXME: Virtual files can't actually model the EOF position correctly, so
-  // if this virtual file would start at EOF, just hope the physical location
-  // will do.
-  if (startOfLine != swiftSourceManager.getRangeForBuffer(mirrorID).getEnd()) {
-    bool isNewVirtualFile = swiftSourceManager.openVirtualFile(
-        startOfLine, presumedFile, presumedLoc.getLine() - bufferLineNumber);
-    if (isNewVirtualFile) {
-      SourceLoc endOfLine = findEndOfLine(swiftSourceManager, loc, mirrorID);
-      swiftSourceManager.closeVirtualFile(endOfLine);
-    }
+  // TODO: The LineTable is obstensibly part of the "internals", should we be
+  //       relying on it here?
```

Comment:
`entries()` is a function I've added on the LLVM side so I'd need to merge that before this one, but I'd like some opinions on this before doing that.

I'd like to avoid creating a virtual file for every line within a line directive. The line table should be complete in all cases other than diagnostics, so this seemed like the easiest option - but if anyone has a better idea I'd be happy to implement that instead :)